### PR TITLE
target_manager: handle all file filtering in python instead of ocaml

### PR DESCRIPTION
**Pull request by brendongo**
This PR adds logic in python to manage which files to run semgrep rules on.

- Respects per rule file ignores and only runs semgrep-core on necessary files instead of post-filtering
- Regex and semgrep-core rules share logic on what files to run on
- Code in semgrep-core that handles file include/exclude was not removed
- Adds support for respecting gitignore (for now turned off. we can turn on and add a flag in a follow up PR)
- Allows cli include/exclude behavior to mimic rule path config style so we don't have two separate include/exclude syntaxes but will expose the cli changes in a follow up PR
- passing list of files to semgrep-core as a file will also do in a follow up PR

### CODE REVIEW

File: semgrep/semgrep/target_manager.py [hunk 1]
```diff
@@ -0,0 +1,223 @@
...
+        """
+        if not includes:
+            return arr
+
+        return set(elem for elem in arr if TargetManager.match_glob(elem, includes))
+
+    @staticmethod
+    def filter_excludes(arr: Set[Path], excludes: List[str]) -> Set[Path]:
+        """
+            Returns all elements in arr that do not match any excludes excludes
```
Comment (minusworld) [inline semgrep/semgrep/target_manager.py:187]: "excludes" here twice ✌️

**Review by minusworld — APPROVED** [review]
looks good to me

**Comment by rcoh** [comment]
FYI, this incurs a 4x performance regression on the sentry repo. (on top of another recent 2x regression)

**Comment by brendongo** [comment]
It does look like we get a speed up in almost all the other perf tests... which is unexpected...

**Comment by rcoh** [comment]
by number, most tests are actually testing non-language (eg. Python against a repo of Java files). my guess is that's what you're seeing?

**Comment by brendongo** [comment]
Looks like caching AST files makes the performance regression for sentry only a 50% increase instead of a 400% which seems reasonable to me. Will turn on caching in a separate PR

**Comment by DrewDennison** [comment]
why are we not doing something that is similar to https://github.com/psf/black/pull/878/files ?

File: semgrep/semgrep/core_runner.py [hunk 2]
```diff
@@ -273,7 +260,7 @@ def _run_rules(
                     if equivalences:
                         cmd += ["-equivalences", equiv_fout.name]
                     cmd += ["-j", str(self._jobs)]
-                    cmd += [*self.targeting_options, *[str(path) for path in targets]]
+                    cmd += [str(path) for path in targets]
```
Comment (DrewDennison) [inline semgrep/semgrep/core_runner.py:263]: this will break on large repos

File: semgrep/semgrep/target_manager.py [hunk 3]
```diff
@@ -0,0 +1,234 @@
...
+
+
+def lang_to_exts(language: str) -> List[str]:
+    """
+        Convert language to expected file extensions
+
+        If language is not a supported semgrep language then
+        raises UnknownLanguageError
+    """
+    if language in ["python", "python2", "python3", "py"]:
```
Comment (DrewDennison) [inline semgrep/semgrep/target_manager.py:19]: is this logic duplicated in ocaml or hopefully this is now the single source of truth?

**Comment by brendongo** [comment]
I don't think that change in black supports nested gitignores and files that match gitignore but a user explicitly tracks on git anyway (added before the ignore or git add -f). Really going to `git ls-files` is much less error prone since it will match exactly what git sees.

File: semgrep/semgrep/target_manager.py [hunk 4]
```diff
@@ -0,0 +1,234 @@
...
+        """
+        extensions = lang_to_exts(language)
+        expanded: Set[Path] = set()
+
+        for ext in extensions:
+            if visible_to_git_only:
+                try:
+                    # Tracked files
+                    tracked_output = subprocess.check_output(
+                        ["git", "ls-files", f"*.{ext}"],
```
Comment (DrewDennison) [inline semgrep/semgrep/target_manager.py:97]: what's the run time of this command? Seems slow compared to extracting globs from .gitignore

File: semgrep/semgrep/target_manager.py [hunk 5]
```diff
@@ -0,0 +1,234 @@
...
+                untracked_unignored = TargetManager._parse_output(
+                    untracked_output, curr_dir
+                )
+
+                expanded = expanded.union(tracked)
+                expanded = expanded.union(untracked_unignored)
+
+            else:
+                output = subprocess.run(
+                    ["find", curr_dir, "-type", "f", "-name", f"*.{ext}"],
```
Comment (DrewDennison) [inline semgrep/semgrep/target_manager.py:131]: shouldn't we be using `Path(.).rglob("*")`?

**Review by DrewDennison — DISMISSED** [review]

File: semgrep/semgrep/target_manager.py (hunk 5 above)
Comment (brendongo) [reply semgrep/semgrep/target_manager.py:131]: This is exactly what semgrep-core does right now

File: semgrep/semgrep/target_manager.py (hunk 4 above)
Comment (brendongo) [reply semgrep/semgrep/target_manager.py:97]: `find . -type f -name "*.py"  0.00s user 0.00s system 87% cpu 0.007 total`
`git ls-files "*.py"  0.00s user 0.01s system 74% cpu 0.014 total`

Negligible difference with find tbh

File: semgrep/semgrep/target_manager.py (hunk 3 above)
Comment (brendongo) [reply semgrep/semgrep/target_manager.py:19]: It still is for now. We can pull out the things in a follow up PR

**Comment by DrewDennison** [comment]
cool maybe we get a bit more perf data and then merge?

File: semgrep/semgrep/target_manager.py (hunk 5 above)
Comment (brendongo) [reply semgrep/semgrep/target_manager.py:131]:
```
python3 -c "from pathlib import Path; [p for p in Path('..').rglob('*.py')]"  0.06s user 0.06s system 96% cpu 0.129 total
python3 -c "from pathlib import Path;"  0.03s user 0.01s system 90% cpu 0.044 total
```

Quick and dirty experiment is using glob takes 10x more time (70 ms vs 7ms) for the small number of files in semgrep

Comment (DrewDennison) [reply semgrep/semgrep/target_manager.py:131]: amazing

Comment (brendongo) [reply semgrep/semgrep/target_manager.py:131]: `time python3 -c 'import subprocess; subprocess.run(["find", "..", "-type", "f", "-name", f"*.py"])'  0.05s user 0.08s system 69% cpu 0.112 total`
With subprocess overhead they're around equivalent.

File: semgrep/semgrep/core_runner.py (hunk 2 above)
Comment (brendongo) [reply semgrep/semgrep/core_runner.py:263]: Yeahp. was thinking of using the -target-file argument i added to semgrep-core in a follow up PR

**Comment by brendongo** [comment]
IMO we will want per rule running of semgrep even with a performance hit just because it offers so much flexibility/control on what things are analyzed and what output to show users (for example it is now possible to show a progress bar) we can hammer out perf differences that remain after AST caching #1002

**Pull request merged**